[torch][amdsmi] Look for amdsmi in ROCM_HOME/ROCM_PATH before using rpath

danzimm opened the pull request:
Summary: ROCm uses ROCM_HOME/ROCM_PATH to specify which version of rocm the user wants to use. This is especially important in multi-version setups. Let's respect that behavior when loading amdsmi.

Test Plan:
CI
```
NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=INIT,COLL MSCCL_ALGO_DIR=~/2fbsource/third-party/rccl/develop/tools/msccl-algorithms RCCL_MSCCLPP_THRESHOLD=(math '128*1024*1024')  RCCL_MSCCLPP_ENABLE=1 ENABLE_MSCCLPP=1 buck2 run fbcode//mode/opt-amd-gpu -m rocm621 fbcode//accelerators/workloads/microbench:bench_comm -- --shape moe_17b --comm_algo nccl_allreduce
```

Differential Revision: D69597647




cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

danzimm commented:
@pytorchbot label "topic: not user facing"

malfet commented on torch/cuda/__init__.py:
```diff
@@ -82,15 +82,22 @@
             class amdsmi_cdll_hook:
                 def __init__(self) -> None:
                     self.original_CDLL = ctypes.CDLL  # type: ignore[misc,assignment]
+                    rocm_home = os.getenv("ROCM_HOME") or os.getenv("ROCM_PATH")
+                    self.paths = (
+                        [os.path.join(rocm_home, "lib/libamd_smi.so")]
+                        if rocm_home
+                        else []
+                    ) + ["libamd_smi.so"]
```
nit (I'm scared of `or` between None and str :)
```suggestion
                    rocm_home = os.getenv("ROCM_HOME", os.getenv("ROCM_PATH", ""))
                    self.paths = [f"{prefix}/{'lib/' if prefix else ''}libamd_smi.so" for prefix in [rocm_home, ""]]
```

danzimm committed:
[torch][amdsmi] Look for amdsmi in ROCM_HOME/ROCM_PATH before using rpath (#147117)

Summary:

ROCm uses ROCM_HOME/ROCM_PATH to specify which version of rocm the user wants to use. This is especially important in multi-version setups. Let's respect that behavior when loading amdsmi.

Test Plan:
CI
```
NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=INIT,COLL MSCCL_ALGO_DIR=~/2fbsource/third-party/rccl/develop/tools/msccl-algorithms RCCL_MSCCLPP_THRESHOLD=(math '128*1024*1024')  RCCL_MSCCLPP_ENABLE=1 ENABLE_MSCCLPP=1 buck2 run fbcode//mode/opt-amd-gpu -m rocm621 fbcode//accelerators/workloads/microbench:bench_comm -- --shape moe_17b --comm_algo nccl_allreduce
```

Reviewed By: malfet, adamomainz

Differential Revision: D69597647
